### PR TITLE
Introduce AggSpec and AggregateAllByTable

### PR DESCRIPTION
**Pull request by devinrsmith**
Partial #1617

### CODE REVIEW

File: qst/graphviz/src/main/java/io/deephaven/graphviz/LabelBuilder.java [hunk 1]
```diff
@@ -135,10 +157,155 @@ public void visit(WhereTable whereTable) {
     }
 
     @Override
-    public void visit(GroupByTable groupByTable) {
-        sb.append("groupBy(");
-        append(Strings::of, groupByTable.columns(), sb);
-        sb.append(')');
+    public void visit(SingleAggregationTable singleAggregationTable) {
```
Comment (rcaudy) [inline qst/graphviz/src/main/java/io/deephaven/graphviz/LabelBuilder.java:160]: Maybe we should just print out the key? I imagine Immutables generates nice toString()s.

File: qst/src/main/java/io/deephaven/qst/TableAdapterImpl.java [hunk 2]
```diff
@@ -235,8 +255,104 @@ public void visit(ReverseAsOfJoinTable raj) {
     }
 
     @Override
-    public void visit(GroupByTable groupByTable) {
-        addOp(groupByTable, parentOps(groupByTable).groupBy(groupByTable.columns()));
+    public void visit(SingleAggregationTable singleAggregationTable) {
```
Comment (rcaudy) [inline qst/src/main/java/io/deephaven/qst/TableAdapterImpl.java:258]: Get rid of this and use aggAllBy.

File: qst/src/main/java/io/deephaven/qst/table/ByTableBase.java [hunk 3]
```diff
@@ -6,5 +6,19 @@
 
 public abstract class ByTableBase extends TableBase implements SingleParentTable {
 
+    public abstract TableSpec parent();
+
     public abstract List<Selectable> columns();
```
Comment (rcaudy) [inline qst/src/main/java/io/deephaven/qst/table/ByTableBase.java:11]: groupByColumns?

File: table-api/src/main/java/io/deephaven/api/agg/Pair.java [hunk 4]
```diff
@@ -5,9 +5,7 @@
 import java.io.Serializable;
 
 /**
- * An aggregation pair represents a {@link #input() input} and {@link #output() output} column for some
- * {@link Aggregation aggregations}. Aggregations that don't have a one-to-one input/output mapping will not need an agg
- * pair.
+ * A pair represents an {@link #input() input} and an {@link #output() output} column for a {@link KeyedAggregation}.
```
Comment (rcaudy) [inline table-api/src/main/java/io/deephaven/api/agg/Pair.java:8]: or {@link KeyedAggregations}.

File: table-api/src/main/java/io/deephaven/api/agg/key/KeyColumnReferences.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    @Override
+    public void visit(KeyPct pct) {
+        out = Collections.emptySet();
+    }
+
+    @Override
+    public void visit(KeySortedFirst sortedFirst) {
+        out = sortedFirst.columns().stream().map(SortColumn::column).collect(Collectors.toSet());
```
Comment (rcaudy) [inline table-api/src/main/java/io/deephaven/api/agg/key/KeyColumnReferences.java:80]: This should come out. Likely column references should become "don't aggregate when single" or something equally silly.

File: table-api/src/main/java/io/deephaven/api/agg/key/KeyPct.java [hunk 6]
```diff
@@ -22,8 +24,10 @@ public boolean averageMedian() {
         return false;
     }
 
-    public Pct withAverage() {
-        return ImmutablePct.builder().percentile(percentile()).pair(pair()).averageMedian(true)
+    public KeyPct withAverage() {
```
Comment (rcaudy) [inline table-api/src/main/java/io/deephaven/api/agg/key/KeyPct.java:27]: Do we want these? I don't care either. Maybe better to have a standard approach,

File: table-api/src/main/java/io/deephaven/api/agg/AggregationOptimizer.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        return optimizer.build();
+    }
+
+    private final LinkedHashMap<Object, List<Pair>> visitOrder = new LinkedHashMap<>();
+
+    public List<Aggregation> build() {
+        List<Aggregation> out = new ArrayList<>();
+        for (Entry<Object, List<Pair>> e : visitOrder.entrySet()) {
+            if (e.getKey() == COUNT_OBJ) {
+                for (Pair pair : e.getValue()) {
```
[the remark this answers is not in the document]
Comment (devinrsmith) [reply table-api/src/main/java/io/deephaven/api/agg/AggregationOptimizer.java:39]: Going to skip for now.

File: table-api/src/main/java/io/deephaven/api/agg/key/KeyEmptyBase.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package io.deephaven.api.agg.key;
+
+public abstract class KeyEmptyBase extends KeyBase {
+
+    @Override
+    public final int hashCode() {
```
[the remark this answers is not in the document]
Comment (devinrsmith) [reply table-api/src/main/java/io/deephaven/api/agg/key/KeyEmptyBase.java:6]: I'm not convinced it would make things faster - even, it may be slower.

**Review by rcaudy — APPROVED** [review]

**Review by rcaudy — APPROVED** [review]

**Pull request merged**